feat(vrl): Add `hmac` function to calculate HMACs (#10333)

sbalmos opened the pull request:
<!--
**Your PR title must conform to the conventional commit spec!**

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/semantic.yml#L20
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

tobz commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,226 @@
+
+    fn parameters(&self) -> &'static [Parameter] {
+        &[
+            Parameter {
+                keyword: "value",
+                kind: kind::ANY,
+                required: true,
+            },
+
+            Parameter {
+                keyword: "key",
+                kind: kind::ANY,
+                required: true,
+            },
+
+            Parameter {
+                keyword: "algorithm",
+                kind: kind::ANY,
+                required: false
+            },
+
+            Parameter {
+                keyword: "encoding",
+                kind: kind::ANY,
+                required: false
+            }
```
Given that you try to convert these the key/value parameters to a bytes representation on lines 18-19 -- where that function basically will either unwrap the value if it's already a bytes value, or throw an error otherwise -- as well as needing the algorithm/encoding to also be bytes... it seems like you should just declare all the parameters as the bytes kind?

tobz reviewed (commented):
From a functional standpoint, this looks good and seems like it follows the recommended approach by the `hmac` crate itself, so that's good. 👍🏻  I left one particular comment about tightening up the parameter declarations, and I'm going to try and get one of the team members more familiar with VRL to take a look to see if there's anything else that can be tightened up.

Additionally, we'll need to document this show it shows up in the website documentation. You can look at `website/cue/reference/remap/functions` to see how we document individual VRL functions.

fuchsnj commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    fn type_def(&self, _: &state::TypeState) -> TypeDef {
+        TypeDef::bytes().fallible()
```
Once the above type changes are made, this function should be infallible. (As long as the inputs are valid, calculating an HMAC can't fail)

sbalmos commented:
Thanks for the comments @tobz and @fuchsnj. I've updated all as you mentioned. When I was writing this, I was basing off of code from sha2(), md5(), and a few others. Not sure why I was using Any-kind to start. Also added in the website documentation cue file. For some reason I thought this came from the returned Examples in the Function impl.

sbalmos commented:
Any possibility of movement on reviewing (and hopefully accepting) this?

fuchsnj commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,226 @@
+    ) -> Compiled {
+        let value = arguments.required("value");
+        let key = arguments.required("key");
+        let algorithm = arguments
```
`algorithm` here can just be kept as `Box<dyn Expression>`. (Since checking the algorithm in `hmac` is necessary anyway, there's no benefit to also trying to check it at compile time here).

Also just a note, but `try_bytes` is not appropriate here. `try_bytes` checks if the _expression_ passed in is a literal `Bytes` expression, but this is called before the value has been resolved / evaluated, so you could have a valid expression that resolves to `Bytes` panic here.

fuchsnj replied:
I would also suggest keeping it an an `Option` instead of applying the default here. The default can be applied in the `hmac` function above. (Otherwise you would need the default to be a boxed expression, rather than just `None`).

fuchsnj commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,226 @@
+            .unwrap_or_else(|| value!("SHA-256"))
+            .try_bytes()
+            .expect("algorithm not bytes");
+        let encoding = arguments
```
Same with encoding, just treat is as optional bytes.

fuchsnj commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,226 @@
+    }
+
+    fn type_def(&self, _: &state::TypeState) -> TypeDef {
+        TypeDef::bytes().infallible()
```
This needs to be `fallible` since the algorithm or encoding chosen could be invalid. However, for the "happy path" where the algorithm and encoding are literals / known at compile-time, this can still be `infallible`.

`Expression::as_value` will try to evaluate an expression at compile-time, and you can check if the value returned from that is valid. (`HmacFn` contains the key and value expression so they can be accessed from the `type_def` function).

fuchsnj commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,226 @@
+
+            Parameter {
+                keyword: "algorithm",
+                kind: kind::BYTES,
```
Just a note: `BYTES` here means any expression that can resolve to `BYTES` (not necessarily just a literal)

fuchsnj reviewed (changes requested):
This mostly looks good. Just some comments about compile-time vs runtime arguments.

sbalmos commented:
Thank you for the comments and suggestions on compile-time vs runtime. Most of the code I have updated as suggested. I kept the defaulting in the `resolve` function so that the context did not have to be further passed down into the `hmac` function. Also, for sake of the new match-default returned Err, and to make the matching cleaner, I slightly changed the `algorithm` and `encoding` resolutions in `hmac` down into `String`s, rather than keeping as `Bytes`.

(separately, the check-spelling PR check is now failing, unrelated to this change... ???)

jsoref commented:
If you merge (or rebase) to master you should be able to resolve any actual issues reported by `check-spelling`.

Sorry, there's a quirk it's tripping on. (Once people base their work on a commit w/ check-spelling, they shouldn't trip on it, and I do hope to figure out what's wrong and fix it in general, although I suspect doing so would require consumers to upgrade to a newer version of check-spelling....)

sbalmos commented:
Done, though now it's failing on my `supersecretkey` in the `Example`s. Not having dealt with this spell checker before, is `super-secret-key` allowed, or something similar?

fuchsnj commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,209 @@
+        hmac(value, key, algorithm, encoding)
+    }
+
+    fn type_def(&self, _: &state::TypeState) -> TypeDef {
```
I think one of my previous comments here was missed: https://github.com/vectordotdev/vector/pull/15371#discussion_r1108976559

The fallibility can be determined based on the inputs. For example, if the inputs are known to be valid at compile-time, this can be marked as infallible.
Example infallible call:
```coffee
hmac("value", "key", "SHA-256", "encoding", "hex")
```

Example fallible call: (value of `algorithm` / `encoding` are unknown)
```coffee
hmac("value", "key", .algorithm, .encoding)
```

fuchsnj reviewed (commented):
Just one change for `type_def`. I think the rest looks good.

jsoref commented:
If you used `super-secret-key`, the spell checker would be totally fine w/ it -- it would see it as `super` + `secret` + `key` each of which are in the dictionary.

sbalmos commented:
The `type_def` part I missed. I've added that, though to be honest this is where I start to get hazy on syntax preferences, etc. The lack of let-chaining is annoying. One thing I think is missing, which I don't know how this would be done if possible, is if there's a compile-time value for `algorithm` or `encoding` that we're checking in `type_def`, and we _know_ that value is invalid (e.g. an `if !valid_static_algo` at line 149), is there a way to compile-time fail? Would that be `TypeDef::never().infallible()`?

sbalmos commented:
Anything else need to be tweaked in this?

fuchsnj commented:
> One thing I think is missing, which I don't know how this would be done if possible, is if there's a compile-time value for `algorithm` or `encoding` that we're checking in `type_def`, and we _know_ that value is invalid (e.g. an `if !valid_static_algo` at line 149), is there a way to compile-time fail? Would that be `TypeDef::never().infallible()`?

This is something that doesn't exist yet, but I would like to see in the future. Right now, it just needs to be `fallible` in that case and it will _always_ fail at runtime.

There is an issue tracking this:
https://github.com/vectordotdev/vector/issues/13968

sbalmos commented:
Ah! Sounds good on the always-fail. And thanks for the fmt run. Let me know if you need anything more before merging.

bruceg commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,243 @@
+        _ => return Err(format!("Invalid algorithm: {}", algorithm).into()),
+    };
+
+    let hash = match encoding.as_str() {
+        "hex" => hex::encode(code_bytes),
+        "base64" => base64::engine::general_purpose::STANDARD.encode(code_bytes),
+        _ => return Err(format!("Invalid encoding: {}", encoding).into()),
+    };
```
Is there a reason this needs to do the encoding within this function instead of just returning bytes and calling `encode_base16` or `encode_base64` on the result?

sbalmos replied:
Mostly opinion here, but it would be an unnecessary, and not readily-apparent, additional step an end-user must take in VRL to separately call `encode_*` on the resulting bytes to get back a usable string. An HMAC is not useful in any other form other than as encoded bytes. Why make this function perform only half the job?

jszwedko replied:
I could be wrong, but I think whether or not the bytes need to be encoded depend on how you are using them? If you are just storing the HMAC somewhere that can handle it as raw bytes, there is no need to encode it.

I tend to agree that we should separate these operations since otherwise we need to remember to add each binary encoding we add to VRL, also to here, and anywhere else we extend this pattern to.  The `encrypt_*` functions are similar in that they just return bytes that can be encoded if desired so not doing it here would at least be consistent within the VRL standard library.

I agree that we should make this behavior clear in the docs though.

sbalmos replied:
At a certain point, I throw in the towel. Most other languages' implementations of HMAC functions I've seen in the wild directly include encode/decode options, and only hex or base64. But I don't hold the commit keys to the kingdom. :) Pushed up another commit removing encoding functionality and made resultant major changes to the function's cue file.

jszwedko replied:
😄 I appreciate the flexibility here. I think returning raw bytes, for now, will maintain consistency within VRL for functions that return bytes, but this is an enhancement we could always add in later if it does turn out to be a common expectation from users (potentially across all functions that return "bytes").

bruceg commented on lib/vrl/stdlib/src/hmac.rs:
```diff
@@ -0,0 +1,243 @@
+    let algorithm = algorithm.try_bytes_utf8_lossy()?.as_ref().to_uppercase();
+    let encoding = encoding.try_bytes_utf8_lossy()?.as_ref().to_lowercase();
+
+    let code_bytes = match algorithm.as_str() {
+        "SHA1" => hmac!(Sha1, key, value),
+        "SHA-224" => hmac!(Sha224, key, value),
+        "SHA-256" => hmac!(Sha256, key, value),
+        "SHA-384" => hmac!(Sha384, key, value),
+        "SHA-512" => hmac!(Sha512, key, value),
+        _ => return Err(format!("Invalid algorithm: {}", algorithm).into()),
+    };
```
Could this table be stored in a look up table that could also generate the `valid_algorithms` used below?

sbalmos replied:
No idea. Not something I want to take on at this point. I had enough fun making the compiler and type system happy by changing to a macro.

bruceg replied:
Fair enough.

bruceg reviewed (commented):
The code here looks reasonable to me, though I do have a question about the utility of integrating the encoding. Ideally, this could also pre-compute the algorithms to use when they are known at compile time, but that's perhaps not going to make a huge performance change.

sbalmos commented:
Well, I'm at a loss now why cue is continuing to complain about incorrect formatting of the `hmac.cue` file. Bone toss here, please?

fuchsnj commented:
@sbalmos you had mixed spaces / tabs on that line. I went ahead and pushed the fix for you.

sbalmos commented:
> @sbalmos you had mixed spaces / tabs on that line. I went ahead and pushed the fix for you.

*VERY* odd, considering I had the file reformatted with `cue fmt` before pushing. Thanks.

sbalmos commented:
Anything else needed or can this be merged? Thanks!

spencergilbert commented:
Sorry for the delay!
